File: app/src/main/java/igrek/forceawaken/service/persistence/AlarmsPersistenceService.java
package igrek.forceawaken.service.persistence;

import android.support.annotation.NonNull;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;

import igrek.forceawaken.domain.alarm.AlarmTrigger;
import igrek.forceawaken.domain.alarm.AlarmsConfig;
import igrek.forceawaken.logger.Logger;
import igrek.forceawaken.logger.LoggerFactory;
import igrek.forceawaken.service.filesystem.InternalDataService;

public class AlarmsPersistenceService {
	
	private Logger logger = LoggerFactory.getLogger();
	private InternalDataService internalDataService;
	
	public AlarmsPersistenceService(InternalDataService internalDataService) {
		this.internalDataService = internalDataService;
	}
	
	public AlarmsConfig readAlarmsConfig() {
		File alarmsConfigFile = getAlarmsConfigFile();
		
		if (!alarmsConfigFile.exists()) {
			logger.warn(alarmsConfigFile.getAbsolutePath() + " file does not exist - creating new AlarmsConfig");
			return new AlarmsConfig();
		}
		
		
		try (FileInputStream fis = new FileInputStream(alarmsConfigFile)) {
			byte fileContent[] = new byte[(int) alarmsConfigFile.length()];
			fis.read(fileContent);
			
			AlarmsConfig unmarshalled = ParcelableUtil.unmarshall(fileContent, AlarmsConfig.CREATOR);
			return unmarshalled;
			
		} catch (Exception e) {
			logger.error(e);
			return null;
		}
		
	}
	
	public void writeAlarmsConfig(AlarmsConfig alarmsConfig) {
		try {
			File alarmsConfigFile = getAlarmsConfigFile();
			FileOutputStream fout = new FileOutputStream(alarmsConfigFile);
			byte[] bytes = ParcelableUtil.marshall(alarmsConfig);
			fout.write(bytes);
			fout.flush();
			fout.close();
		} catch (IOException e) {
			logger.error(e);
		}
	}
	
	@NonNull
	private File getAlarmsConfigFile() {
		File internalDataDir = internalDataService.getInternalDataDir();
		return new File(internalDataDir, "alarmsConfig");
	}
	
	public AlarmsConfig addAlarmTrigger(AlarmTrigger alarmTrigger) {
		AlarmsConfig alarmsConfig = readAlarmsConfig();
		alarmsConfig.getAlarmTriggers().add(alarmTrigger);
		writeAlarmsConfig(alarmsConfig);
		logger.info("Alarm trigger has been added: " + alarmTrigger);
		return alarmsConfig;
	}
	
	public AlarmsConfig removeAlarmTrigger(AlarmTrigger alarmTrigger) {
		AlarmsConfig alarmsConfig = readAlarmsConfig();
		if (alarmsConfig.getAlarmTriggers().remove(alarmTrigger)) {
			writeAlarmsConfig(alarmsConfig);
			logger.info("Alarm trigger has been removed: " + alarmTrigger);
		}
		return alarmsConfig;
	}
}

File: app/src/main/java/igrek/forceawaken/persistence/AlarmsPersistenceService.kt
package igrek.forceawaken.persistence

import igrek.forceawaken.alarm.AlarmTrigger
import igrek.forceawaken.alarm.AlarmsConfig
import igrek.forceawaken.alarm.RepetitiveAlarm
import igrek.forceawaken.info.logger.Logger
import igrek.forceawaken.info.logger.LoggerFactory
import igrek.forceawaken.inject.LazyExtractor
import igrek.forceawaken.inject.LazyInject
import igrek.forceawaken.inject.appFactory
import igrek.forceawaken.system.filesystem.InternalDataService
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

class AlarmsPersistenceService(
        internalDataService: LazyInject<InternalDataService> = appFactory.internalDataService,
) {
    private val internalDataService by LazyExtractor(internalDataService)

    private val logger: Logger = LoggerFactory.logger

    fun readAlarmsConfig(): AlarmsConfig {
        val alarmsConfigFile = alarmsConfigFile
        if (!alarmsConfigFile.exists()) {
            logger.warn(alarmsConfigFile.absolutePath + " file does not exist - creating new AlarmsConfig")
            return AlarmsConfig()
        }
        FileInputStream(alarmsConfigFile).use { fis ->
            val fileContent = ByteArray(alarmsConfigFile.length().toInt())
            fis.read(fileContent)
            val config = ParcelableUtil.unmarshall(fileContent, AlarmsConfig.CREATOR)

            config.alarmTriggers.sortBy { it.triggerTime }
            config.repetitiveAlarms.sortBy { it.triggerTime }

            return config
        }
    }

    private fun writeAlarmsConfig(alarmsConfig: AlarmsConfig) {
        try {
            val alarmsConfigFile = alarmsConfigFile
            val fout = FileOutputStream(alarmsConfigFile)
            val bytes = ParcelableUtil.marshall(alarmsConfig)
            fout.write(bytes)
            fout.flush()
            fout.close()
        } catch (e: IOException) {
            logger.error(e)
        }
    }

    private val alarmsConfigFile: File
        get() {
            val internalDataDir: File = internalDataService.internalDataDir
            return File(internalDataDir, "alarmsConfig")
        }

    fun addAlarmTrigger(alarmTrigger: AlarmTrigger): AlarmsConfig {
        val alarmsConfig: AlarmsConfig = readAlarmsConfig()
        alarmsConfig.alarmTriggers.add(alarmTrigger)
        writeAlarmsConfig(alarmsConfig)
        logger.info("Alarm trigger has been added: $alarmTrigger")
        return alarmsConfig
    }

    fun removeAlarmTrigger(alarmTrigger: AlarmTrigger): AlarmsConfig {
        val alarmsConfig: AlarmsConfig = readAlarmsConfig()
        if (alarmsConfig.alarmTriggers.remove(alarmTrigger)) {
            writeAlarmsConfig(alarmsConfig)
            logger.info("Alarm trigger has been removed: $alarmTrigger")
        }
        return alarmsConfig
    }

    fun addRepetitiveAlarm(repetitiveAlarm: RepetitiveAlarm): AlarmsConfig {
        val alarmsConfig: AlarmsConfig = readAlarmsConfig()
        alarmsConfig.repetitiveAlarms.add(repetitiveAlarm)
        writeAlarmsConfig(alarmsConfig)
        logger.info("Repetitive alarm has been added: $repetitiveAlarm")
        return alarmsConfig
    }

    fun removeRepetitiveAlarm(repetitiveAlarm: RepetitiveAlarm): AlarmsConfig {
        val alarmsConfig: AlarmsConfig = readAlarmsConfig()
        if (alarmsConfig.repetitiveAlarms.remove(repetitiveAlarm)) {
            writeAlarmsConfig(alarmsConfig)
            logger.info("Repetitive alarm has been removed: $repetitiveAlarm")
        }
        return alarmsConfig
    }

    fun updateRepetitiveAlarm(
        repetitiveAlarm: RepetitiveAlarm,
        applyBlock: RepetitiveAlarm.() -> Unit
    ): RepetitiveAlarm {
        val alarmsConfig: AlarmsConfig = readAlarmsConfig()
        val configAlarm = alarmsConfig.repetitiveAlarms.find { it == repetitiveAlarm }
        if (configAlarm == null) {
            logger.error("Cant find alarm in config")
            return repetitiveAlarm
        }
        configAlarm.applyBlock()
        writeAlarmsConfig(alarmsConfig)
        logger.info("Repetitive alarm has been updated: $configAlarm")
        return configAlarm
    }

}